### PR TITLE
tracing defaults for multi and share handles

### PR DESCRIPTION
**Pull request by icing**
Configure a multi handle to provide defaults for added easy handles when it comes to tracing:
- CURLMOPT_DEBUGDATA, CURLMOPT_DEBUGFUNCTION, CURLMOPT_VERBOSE

Configure a shared handle to provide defaults for easy handles that use the share when it comes to tracing:
- CURLSHOPT_DEBUGDATA, CURLSHOPT_DEBUGFUNCTION, CURLSHOPT_VERBOSE

Defaults for DEBUGDATA and DEBUGFUNCTION apply to easy handles where none of them are already set.
Default to VERBOSE act only as enabler. It will never turn off verbosity in an easy handle that has it set.

This is now used in the `curl` tool itself.

- when using curl verbose, the closing of connections in the pool become visible
- setting tracing on the easy handles could be removed, however `--libcurl` would then no longer generate the code for it. That is why I left it in.

### CODE REVIEW

File: docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3 [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+CURLcode curl_multi_setopt(CURLM *handle, CURLOPT_DEBUGFUNCTION,
+                          debug_callback);
+.SH DESCRIPTION
+Pass a pointer to your callback function, which should match the prototype
+that is defined for \fICURLOPT_DEBUGFUNCTION(3)\fP.
+
+\fICURLMOPT_DEBUGFUNCTION(3)\fP and \fICURLMOPT_DEBUGDATA\fP can be used
+as defaults for all easy handles that are added to the multi. They are also
+used for internal handles, for examples when closing cached connections.
+They take effect only for easy handles added to the multi handle afterwards.
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3:41]: I did not get this when I first read it. I propose adding a few more words:

```suggestion
These options take effect and are used only for the easy handles that are added to the multi handle after these options have been set for the multi handle.
```

File: docs/libcurl/opts/CURLMOPT_VERBOSE.3 [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+.fi
+.SH DESCRIPTION
+Set the \fIonoff\fP parameter to 1 to make the library display a lot of
+verbose information about its operations on \fIhandle\fPs added to
+this multi.
+
+\fICURLMOPT_VERBOSE(3)\fP can be used as default for all easy handles
+that are added to the multi. They are also used for internal handles,
+for examples when closing cached connections.
+They take effect only for easy handles added to the multi handle afterwards.
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_VERBOSE.3:42]:
```suggestion
This option take effect and is used only for the easy handles that are added to the multi handle after this option is set for the multi handle.
```

File: docs/libcurl/opts/CURLMOPT_VERBOSE.3 [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+.\" **************************************************************************
+.\"
+.TH CURLMOPT_VERBOSE 3 "17 Oct 2023" libcurl libcurl
+.SH NAME
+CURLMOPT_VERBOSE \- verbose mode
+.SH SYNOPSIS
+.nf
+#include <curl/curl.h>
+
+CURLcode curl_multi_setopt(CURLM *handle, CURLOPT_VERBOSE, long onoff);
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_VERBOSE.3:32]:
```suggestion
CURLcode curl_multi_setopt(CURLM *handle, CURLMOPT_VERBOSE, long onoff);
```

File: docs/libcurl/opts/CURLMOPT_VERBOSE.3 [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+that are added to the multi. They are also used for internal handles,
+for examples when closing cached connections.
+They take effect only for easy handles added to the multi handle afterwards.
+.SH PROTOCOLS
+All
+.SH EXAMPLE
+.nf
+main()
+{
+  GlobalInfo setup;
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_VERBOSE.3:49]: Maybe drop this line?

File: docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3 [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+NULL
+.SH PROTOCOLS
+All
+.SH EXAMPLE
+.nf
+main()
+{
+  struct data my_tracedata;
+
+  GlobalInfo setup;
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3:55]: Maybe remove this?

File: docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3 [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+  /* ... use debug callback and custom pointer */
+  curl_multi_setopt(multi, CURLMOPT_DEBUGFUNCTION, my_trace);
+  curl_multi_setopt(multi, CURLMOPT_DEBUGDATA, &my_tracedata);
+}
+.fi
+.SH AVAILABILITY
+Always
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3:62]:
```suggestion
This option was added in libcurl 8.5.0.
```

File: docs/libcurl/opts/CURLMOPT_VERBOSE.3 [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  /* ... use debug callback and custom pointer */
+  curl_multi_setopt(multi, CURLMOPT_VERBOSE, 1L);
+}
+.fi
+.SH AVAILABILITY
+Always
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_VERBOSE.3:55]:
```suggestion
This option was added in libcurl 8.5.0.
```

File: docs/libcurl/opts/CURLSHOPT_DEBUGDATA.3 [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+{
+  struct data my_tracedata;
+
+  share = curl_share_init();
+  curl_share_setopt(share, CURLSHOPT_DEBUGFUNCTION, my_trace);
+  curl_share_setopt(share, CURLSHOPT_DEBUGDATA, &my_tracedata);
+}
+.fi
+.SH AVAILABILITY
+Always
```
Comment (bagder) [inline docs/libcurl/opts/CURLSHOPT_DEBUGDATA.3:64]:
```suggestion
This option was added in libcurl 8.5.0.
```

File: docs/libcurl/opts/CURLMOPT_DEBUGDATA.3 [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  struct data my_tracedata;
+
+  GlobalInfo setup;
+  /* ... use debug callback and custom pointer */
+  curl_multi_setopt(multi, CURLMOPT_DEBUGFUNCTION, my_trace);
+  curl_multi_setopt(multi, CURLMOPT_DEBUGDATA, &my_tracedata);
+}
+.fi
+.SH AVAILABILITY
+Always
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_DEBUGDATA.3:63]:
```suggestion
This option was added in libcurl 8.5.0.
```

File: docs/libcurl/opts/CURLSHOPT_VERBOSE.3 [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+main()
+{
+  struct data my_tracedata;
+
+  share = curl_share_init();
+  curl_share_setopt(share, CURLSHOPT_VERBOSE, 1L);
+}
+.fi
+.SH AVAILABILITY
+Always
```
Comment (bagder) [inline docs/libcurl/opts/CURLSHOPT_VERBOSE.3:56]:
```suggestion
This option was added in libcurl 8.5.0.
```

File: docs/libcurl/opts/CURLSHOPT_DEBUGFUNCTION.3 [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+{
+  struct data my_tracedata;
+
+  share = curl_share_init();
+  curl_share_setopt(share, CURLSHOPT_DEBUGFUNCTION, my_trace);
+  curl_share_setopt(share, CURLSHOPT_DEBUGDATA, &my_tracedata);
+}
+.fi
+.SH AVAILABILITY
+Always
```
Comment (bagder) [inline docs/libcurl/opts/CURLSHOPT_DEBUGFUNCTION.3:62]:
```suggestion
This option was added in libcurl 8.5.0.
```

File: docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3 (hunk 1 above)
Comment (icing) [reply docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3:41]: ack

File: docs/libcurl/opts/CURLMOPT_VERBOSE.3 (hunk 2 above)
Comment (icing) [reply docs/libcurl/opts/CURLMOPT_VERBOSE.3:42]: ack

File: docs/libcurl/opts/CURLMOPT_VERBOSE.3 (hunk 3 above)
Comment (icing) [reply docs/libcurl/opts/CURLMOPT_VERBOSE.3:32]: ack

File: docs/libcurl/opts/CURLMOPT_VERBOSE.3 (hunk 4 above)
Comment (icing) [reply docs/libcurl/opts/CURLMOPT_VERBOSE.3:49]: ack

File: docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3 (hunk 5 above)
Comment (icing) [reply docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3:55]: ack

File: docs/libcurl/opts/CURLMOPT_VERBOSE.3 (hunk 7 above)
Comment (icing) [reply docs/libcurl/opts/CURLMOPT_VERBOSE.3:55]: ack

File: docs/libcurl/opts/CURLSHOPT_DEBUGDATA.3 (hunk 8 above)
Comment (icing) [reply docs/libcurl/opts/CURLSHOPT_DEBUGDATA.3:64]: ack

File: docs/libcurl/opts/CURLMOPT_DEBUGDATA.3 (hunk 9 above)
Comment (icing) [reply docs/libcurl/opts/CURLMOPT_DEBUGDATA.3:63]: ack

File: docs/libcurl/opts/CURLSHOPT_VERBOSE.3 (hunk 10 above)
Comment (icing) [reply docs/libcurl/opts/CURLSHOPT_VERBOSE.3:56]: ack

File: docs/libcurl/opts/CURLSHOPT_DEBUGFUNCTION.3 (hunk 11 above)
Comment (icing) [reply docs/libcurl/opts/CURLSHOPT_DEBUGFUNCTION.3:62]: ack

File: docs/libcurl/opts/CURLMOPT_DEBUGDATA.3 [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+CURLcode curl_multi_setopt(CURLM *handle, CURLMOPT_DEBUGDATA, void *pointer);
+.fi
+.SH DESCRIPTION
+Pass a \fIpointer\fP to whatever you want passed in to your
+\fICURLMOPT_DEBUGFUNCTION(3)\fP in the last void * argument. This pointer is
+not used by libcurl, it is only passed to the callback.
+
+\fICURLMOPT_DEBUGFUNCTION(3)\fP and \fICURLMOPT_DEBUGDATA\fP can be used
+as defaults for all easy handles that are added to the multi. They are also
+used for internal handles, for examples when closing cached connections.
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_DEBUGDATA.3:41]:
```suggestion
used for internal handles, for example when closing cached connections.
```

File: docs/libcurl/opts/CURLMOPT_DEBUGDATA.3 [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+NULL
+.SH PROTOCOLS
+All
+.SH EXAMPLE
+.nf
+main()
+{
+  struct data my_tracedata;
+
+  GlobalInfo setup;
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_DEBUGDATA.3:58]:
```suggestion
```

File: docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3 [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+CURLcode curl_multi_setopt(CURLM *handle, CURLOPT_DEBUGFUNCTION,
+                          debug_callback);
+.SH DESCRIPTION
+Pass a pointer to your callback function, which should match the prototype
+that is defined for \fICURLOPT_DEBUGFUNCTION(3)\fP.
+
+\fICURLMOPT_DEBUGFUNCTION(3)\fP and \fICURLMOPT_DEBUGDATA\fP can be used
+as defaults for all easy handles that are added to the multi. They are also
+used for internal handles, for examples when closing cached connections.
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3:40]:
```suggestion
used for internal handles, for example when closing cached connections.
```

File: docs/libcurl/opts/CURLMOPT_VERBOSE.3 [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+CURLcode curl_multi_setopt(CURLM *handle, CURLMOPT_VERBOSE, long onoff);
+.fi
+.SH DESCRIPTION
+Set the \fIonoff\fP parameter to 1 to make the library display a lot of
+verbose information about its operations on \fIhandle\fPs added to
+this multi.
+
+\fICURLMOPT_VERBOSE(3)\fP can be used as default for all easy handles
+that are added to the multi. They are also used for internal handles,
+for examples when closing cached connections.
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_VERBOSE.3:41]:
```suggestion
for example when closing cached connections.
```

File: docs/libcurl/opts/CURLMOPT_VERBOSE.3 [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+This option takes effect and is used only for the easy handles that are
+added to the multi handle after this option is set for the multi handle.
+.SH PROTOCOLS
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_VERBOSE.3:44]:
```suggestion
.SH DEFAULT
This is treated as if set to 0 by default.
.SH PROTOCOLS
```

File: docs/libcurl/opts/CURLSHOPT_DEBUGDATA.3 [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+                             void *pointer);
+.fi
+.SH DESCRIPTION
+Pass a \fIpointer\fP to whatever you want passed in to your
+\fICURLSHOPT_DEBUGFUNCTION(3)\fP in the last void * argument. This pointer is
+not used by libcurl, it is only passed to the callback.
+
+\fICURLSHOPT_DEBUGFUNCTION(3)\fP and \fICURLSHOPT_DEBUGDATA\fP can be used
+as defaults for all easy handles that use the same share. They are also
+used for internal handles, for examples when closing cached connections.
```
Comment (bagder) [inline docs/libcurl/opts/CURLSHOPT_DEBUGDATA.3:42]:
```suggestion
used for internal handles, for example when closing cached connections.
```

File: docs/libcurl/opts/CURLSHOPT_DEBUGFUNCTION.3 [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+
+CURLSHcode curl_share_setopt(CURLSH *share, CURLSHOPT_DEBUGFUNCTION,
+                             debug_callback);
+.SH DESCRIPTION
+Pass a pointer to your callback function, which should match the prototype
+that is defined for \fICURLOPT_DEBUGFUNCTION(3)\fP.
+
+\fICURLSHOPT_DEBUGFUNCTION(3)\fP and \fICURLSHOPT_DEBUGDATA\fP can be used
+as defaults for all easy handles that use the same share. They are also
+used for internal handles, for examples when closing cached connections.
```
Comment (bagder) [inline docs/libcurl/opts/CURLSHOPT_DEBUGFUNCTION.3:40]:
```suggestion
used for internal handles, for example when closing cached connections.
```

File: docs/libcurl/opts/CURLSHOPT_VERBOSE.3 [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+                             long onoff);
+.fi
+.SH DESCRIPTION
+Set the \fIonoff\fP parameter to 1 to make the library display a lot of
+verbose information about its operations on \fIhandle\fPs used with
+this share.
+
+\fICURLSHOPT_VERBOSE(3)\fP can be used as default for all easy handles
+that use this share via \fICURLOPT_SHARE(3)\fP. It is also used for
+internal handles, for examples when closing cached connections.
```
Comment (bagder) [inline docs/libcurl/opts/CURLSHOPT_VERBOSE.3:42]:
```suggestion
internal handles, for example when closing cached connections.
```

File: docs/libcurl/opts/CURLSHOPT_VERBOSE.3 [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+.SH PROTOCOLS
```
Comment (bagder) [inline docs/libcurl/opts/CURLSHOPT_VERBOSE.3:43]:
```suggestion
.SH DEFAULT
0
.SH PROTOCOLS
```

File: docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3 [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+These options take effect and are used only for the easy handles that are
+added to the multi handle after these options have been set for the multi
+handle.
+
+Values have the same restrictions and use patterns as defined for
+\fICURLOPT_DEBUGFUNCTION(3)\fP and \fICURLOPT_DEBUGDATA(3)\fP.
```
Comment (bagder) [inline docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3:46]:
```suggestion
\fICURLOPT_DEBUGFUNCTION(3)\fP and \fICURLOPT_DEBUGDATA(3)\fP.

Setting this option to NULL makes the function not be used from the share object and instead the value set in the easy handle dictates what function to call.
```

File: docs/libcurl/opts/CURLSHOPT_DEBUGFUNCTION.3 [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+They take effect only for easy handles when this share is set for them
+using \fICURLOPT_SHARE(3)\fP.
+
+Values have the same restrictions and use patterns as defined for
+\fICURLOPT_DEBUGFUNCTION(3)\fP and \fICURLOPT_DEBUGDATA(3)\fP.
```
Comment (bagder) [inline docs/libcurl/opts/CURLSHOPT_DEBUGFUNCTION.3:45]:
```suggestion
\fICURLOPT_DEBUGFUNCTION(3)\fP and \fICURLOPT_DEBUGDATA(3)\fP.

Setting this option to NULL makes the debug function not be used from the multi object and instead the value set in the easy handle dictates what function to call.
```

**Review by bagder — COMMENTED** [review]
I went over it again and I have some polish to propose!

**Comment by jay** [comment]
I've added CURLSHOPT_STDERR and CURLMOPT_STDERR, which is used by verbose, although I have to wonder if all these extra options are overkill and maybe there is a better way to solve this like a CURLOPT_DEBUGINHERIT or something so that the caller knows their CURLOPT_STDERR and CURLOPT_DEBUGFUNCTION/DATA can be used after the easy handle is cleaned up.

File: docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3 (hunk 21 above)
Comment (icing) [reply docs/libcurl/opts/CURLMOPT_DEBUGFUNCTION.3:46]: Both options would need to be set to NULL in order to not override the ease settings. I'll adapt the suggested wording.

File: docs/libcurl/opts/CURLSHOPT_DEBUGFUNCTION.3 (hunk 22 above)
Comment (icing) [reply docs/libcurl/opts/CURLSHOPT_DEBUGFUNCTION.3:45]: see above

**Comment by icing** [comment]
Jay, you changed the way these variables are now inherited, seemingly in a way not intended by me. Also, the test cases hard crash for me.

Not a happy Monday morning...

**Comment by jay** [comment]
> Jay, you changed the way these variables are now inherited, seemingly in a way not intended by me. Also, the test cases hard crash for me.

It's a WIP. How did you intend? My changes were in line with this:

> Defaults for DEBUGDATA and DEBUGFUNCTION apply to easy handles where none of them are already set.
> Default to VERBOSE act only as enabler. It will never turn off verbosity in an easy handle that has it set.

**Comment by icing** [comment]
Had to force push my latest changes. Thanks for addressing the NULL-deref in the meantime.

**Comment by jay** [comment]
I think the inheritance issue complicates things. If someone sets easy handle stderr, and not easy handle debug function, and sets multi debug function, then it would make sense to for multi debug function to be inherited but if I'm reading this right then right now that doesn't happen.

In the next few days I am going to write up an alternate proposal to do a CURLOPT_DEBUGINTERNAL to set the easy handle's debug settings on the conn cache handle and we can see how that looks.

**Comment by icing** [comment]
> I think the inheritance issue complicates things. If someone sets easy handle stderr, and not easy handle debug function, and sets multi debug function, then it would make sense to for multi debug function to be inherited but if I'm reading this right then that doesn't happen.

The design in this PR is intended to make the full set of DEBUG/VERBOSE inheritable only. This is intended to give the "least surprise" to code that is setting these properties on easy handles. They will function as before.
 
> In the next few days I am going to write up an alternate proposal to do a CURLOPT_DEBUGINTERNAL to set the easy handle's debug settings on the conn cache handle and we can see how that looks.

The thing you seem to have in mind, and pushed into this PR before I reverted, follows another idea. Namely, that each individual property should be inherited. This gives rise to the problem of how to manage the case when the application does not want a property to be inherited, e.g. have DEBUGFUNCTION not set on an easy handle.

I find that leads to more "complicated" management and therefore decided not to do that. Looking forward to your proposal on how to manage this.

**Comment by icing** [comment]
Did not make the cut. We are still looking for the optimal approach.

**Pull request closed without merging**